lib/runner/ssh-tunnel-agent.js

```diff
@@ -0,0 +1,221 @@
+            self.isReady(false);
+        });
+
+        sshc.on('ready', function () {
+            var update = {config: sshConfig, client: sshc};
+
+            !_.includes(self.mapping, update) && self.mapping.push(update);
+            self.isReady(true);
+        });
+
+        return sshc;
+    },
+
+    // Helper function to modify proxies to listen on local port.
+    setLocalProxy: function(scope, localport) {
+        this.proxyConfigList.all().forEach(function (pc) {
+            pc.host = LOCALHOST;
+            pc.port = localport;
+            scope.options.proxies.upsert(pc);
+        });
+    },
+
+    // Helper function to destroy streams once they are closed.
+    closeSock: function(sock, stream) {
+        sock.destroy();
+        stream.destroy();
+    },
+
+    /**
+     * Forwards a messages on to some another configured port on ssh host. There could be a http proxy running on '
+     * this port, or maybe something else in any case the traffic would be forearded to this port once it
+     * reacher ssh host.
+     *
+     * @param {SshClient} sshClient
+     * @param {SshConfig} sshConfig
+     * @param {Socket} sock
+     */
+    forward: function (sshClient, sshConfig, sock) {
+        var self = this,
+            host = sshConfig.auth.host;
+
+        sshClient.forwardOut('', sock.localPort, sshConfig.auth.host, sshConfig.proxyPort, function(err, stream) {
+            if (err) {
+                self.reporter.emit('tunnelError', new Error('request cannot be tunneled to host: ' + host +
+                        ', ' + err.message));
+                self.closeSock(sock, stream);
+
+                return;
+            }
+
+            sock.pipe(stream).pipe(sock);
+
+            sock.on('error', function(err) {
+                self.reporter.emit('tunnelError', new Error('connection was closed with errors for tunnel host: ' +
+                        host + ', ' + err.message));
+                self.closeSock(sock, stream);
+            });
+
+            stream.on('error', function(err) {
+                self.reporter.emit('tunnelError', new Error('connection was closed with errors for tunnel host: ' +
+                        host + ', ') + err.message);
+                self.closeSock(sock, stream);
+            });
+        });
+    },
+
+    /**
+     * Given a proxy list creates and starts different ssh tunnels based on proxy list.
+     *
+     * @param {ProxyConfigList} proxyConfigList list of proxies
+     * @param {Number} localport port listening locally for requests and forwarding them through tunnel
+     * @param {Object} scope
+     * @param {Function} cb callback function that will be called once all the tunnels are
+     * initialised or some error occurred
+     *
+     */
+
+    start: function (proxyConfigList, localport, scope, cb) {
+        var self = this,
+            sshConfigList;
+
+
+        self.proxyConfigList = (ProxyConfigList.isProxyConfigList(proxyConfigList)) ?
```

Comment:
Shift sanitization to a different statement above, perform only assignment here.